src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala

```diff
@@ -0,0 +1,130 @@
+package is.hail.expr.ir.functions
+
+import is.hail.expr.ir._
+import is.hail.expr.types._
+import is.hail.expr.types.coerce
+
+object ArrayFunctions extends RegistryFunctions {
+
+  def registerAll() {
+    registerIR("size", TArray(tv("T")))(ArrayLen)
+
+    val arrayOps: Array[(String, (IR, IR) => IR)] =
+      Array(
+        ("*", ApplyBinaryPrimOp(Multiply(), _, _)),
+        ("/", ApplyBinaryPrimOp(FloatingPointDivide(), _, _)),
+        ("//", ApplyBinaryPrimOp(RoundToNegInfDivide(), _, _)),
+        ("+", ApplyBinaryPrimOp(Add(), _, _)),
+        ("-", ApplyBinaryPrimOp(Subtract(), _, _)),
+        ("**", { (ir1: IR, ir2: IR) => Apply("**", Seq(ir1,ir2)) }),
+        ("%", { (ir1: IR, ir2: IR) => Apply("%", Seq(ir1,ir2)) }))
+
+    for ((stringOp, irOp) <- arrayOps) {
+      registerIR(stringOp, TArray(tnum("T")), tv("T")) { (a, c) =>
+        val i = genUID()
+        ArrayMap(a, i, irOp(Ref(i, c.typ), c))
+      }
+
+      registerIR(stringOp, tnum("T"), TArray(tv("T"))) { (c, a) =>
```

Comment:
Can you double check all of the tests in `test_expr` such as `test_modulus`, `test_addition`, etc. go through this new path? Also, I don't see any tests for where both operands are arrays in the Python tests. Can you add them?

Reply:
I think the array, array ops are already being tested? e.g. in `test_modulus` I see

```
        self.check_expr(a_int32 % int32_3s, expected, tarray(tint32))
        self.check_expr(a_int64 % int32_3s, expected, tarray(tint64))
        self.check_expr(a_float32 % int32_3s, expected, tarray(tfloat32))
        self.check_expr(a_float64 % int32_3s, expected, tarray(tfloat64))
```

Reply:
( also checked that there's no `AST` warnings in the log when these tests run)

Reply:
`int32_3s` is an array. The cast gets handled in python.